### DISCUSSION TITLE
How to create different distribution packages for different projects

### DISCUSSION THREAD

**Discussion by johnsonlien** [mesonbuild/meson#11612] [Q&A]
Hi!

**Background**
I have a monolithic repo that multiple programs use and update. We defined all dependencies, libraries, and executables in multiple subdirectories of the project. Each distribution follows a similar structure
```
program_root
    a_dir
        application (for each application)
            bin
            configuration_files
            lib
            scripts
    b_dir (same structure as a_dir)
```    

**Current solution to create release**
To create a program's specific distribution package, the different teams wrote a bash script to, for each needed component, they run the following steps:

1. turn off all build options
2. turn only the build options required for that specific component and redefine a distribution folder location
3. rebuild the project

**Question**
Rather than use a script create our distribution package, is there a way to utilize meson to do it instead? I'm trying to make adding future program distributions easier to create since the person writing the script can mistakenly miss setting a required meson option (or turn on an option that isn't needed) and hopefully utilize the current `build` directory so we don't have to always rebuild for each component.

**Comment by xclaesse** [comment]
If you turn each programs into their own subproject, you can then "meson dist" them separately even when there are in the same git repo. That means you need to structure your tree like: subprojects/a_dir, subprojects/b_dir, etc, and the meson.build at the root would do
```meson
subproject('a_dir')
subproject('b_dir')
```
